### PR TITLE
Add Resiliency Feature

### PR DESCRIPTION
**Pull request by jasoncoolmax**
(For #643) To achieve uninterrupted services access even during an Evo STS outage, server may return ext_expires_in value to indicate the extended lifetime of an access token.

The following changes are made to support the resiliency feature:

1) In ADAuthenticationContext,
Property **extendedLifetimeEnabled** is added for developers to opt in the feature.

2) In ADAuthenticationResult,
Property **extendedLifeTimeToken** is added to indicate whether the token being returned is a token only valid in terms of extended lifetime.

3) In ADAcquireTokenSilentHandler, _(token return logic)_
Through out the normal cache lookup sequence(getAccessToken, tryRT, tryMRRT, etc.), we store the valid access token in terms of ext_expires_in (if we find any) to a variable **_extendedLifetimeAccessTokenItem**. In the end of the lookup sequence, we will return **_extendedLifetimeAccessTokenItem** as result if no valid AT is found, *extendedLifetimeEnabled* is on and also server is unavailable.

4) In ADAcquireTokenSilentHandler, _(token delete logic)_
An access token will be deleted only if the token expires in terms both _expires_in_ and _ext_expires_in_.

5) In ADWebAuthRequest,
Now it will retry once after a second if server response is 500/503/504. (previously only retry for 500/503)

6) Unit tests added.

### CODE REVIEW

File: ADAL/src/cache/ADTokenCacheItem+Internal.h [hunk 1]
```diff
@@ -59,4 +59,7 @@
              level:(ADAL_LOG_LEVEL)level
      correlationId:(NSUUID*)correlationId;
 
+/*! Return NO only if the item contains an access token and ext_expires_in in additionalServer has not expired. */
```
Comment (heidijinxujia) [inline ADAL/src/cache/ADTokenCacheItem+Internal.h]: rename additionalServer

File: ADAL/src/cache/ADTokenCacheItem+Internal.m [hunk 2]
```diff
@@ -214,6 +214,9 @@ - (BOOL)fillItemWithResponse:(NSDictionary*)responseDictionary
     [additional removeObjectForKey:@"expires_in"];
     [additional removeObjectForKey:@"expires_on"];
     
+    // reformat ext_expires_in to ext_expires_on in the dictionary
+    [self reformatExtendedExpiration:additional];
```
Comment (heidijinxujia) [inline ADAL/src/cache/ADTokenCacheItem+Internal.m]: rename it or use existing funciton.

File: ADAL/src/cache/ADTokenCacheItem+Internal.m [hunk 3]
```diff
@@ -326,5 +329,27 @@ - (void)setAdditionalClient:(NSMutableDictionary *)additionalClient
     SAFE_ARC_RETAIN(_additionalClient);
 }
 
+- (void)reformatExtendedExpiration:(NSMutableDictionary *)additionalDictionary
```
Comment (RPangrle) [inline ADAL/src/cache/ADTokenCacheItem+Internal.m]: Move this logic into -[ADTokenCacheItem fillExpiration:]

File: ADAL/src/cache/ADTokenCacheItem+Internal.m [hunk 4]
```diff
@@ -326,5 +329,27 @@ - (void)setAdditionalClient:(NSMutableDictionary *)additionalClient
...
+                                 forKey:@"ext_expires_on"];
+    }
+}
+
+- (BOOL)isExtendedLifetimeExpired
+{
+    NSDate* extendedExpiresOn = [_additionalServer valueForKey:@"ext_expires_on"];
+    
+    //extended lifetime is only valid if it contains an access token
+    if ([NSString adIsStringNilOrBlank:_accessToken] || !extendedExpiresOn)
```
Comment (heidijinxujia) [inline ADAL/src/cache/ADTokenCacheItem+Internal.m]: remove first condition

File: ADAL/src/cache/ADTokenCacheItem+Internal.m [hunk 5]
```diff
@@ -326,5 +329,27 @@ - (void)setAdditionalClient:(NSMutableDictionary *)additionalClient
...
+{
+    NSString* extendedExpiresIn = [additionalDictionary valueForKey:@"ext_expires_in"];
+    if (![NSString adIsStringNilOrBlank:extendedExpiresIn] && [extendedExpiresIn respondsToSelector:@selector(doubleValue)])
+    {
+        [additionalDictionary setObject:[NSDate dateWithTimeIntervalSinceNow:[extendedExpiresIn doubleValue]]
+                                 forKey:@"ext_expires_on"];
+    }
+}
+
+- (BOOL)isExtendedLifetimeExpired
```
Comment (heidijinxujia) [inline ADAL/src/cache/ADTokenCacheItem+Internal.m]: isExtendedLifetimeValid?

File: ADAL/src/cache/ADTokenCacheItem+Internal.m [hunk 6]
```diff
@@ -326,5 +329,27 @@ - (void)setAdditionalClient:(NSMutableDictionary *)additionalClient
...
+{
+    NSDate* extendedExpiresOn = [_additionalServer valueForKey:@"ext_expires_on"];
+
```
Comment (RPangrle) [inline ADAL/src/cache/ADTokenCacheItem+Internal.m]: nil check this before you use it.

File: ADAL/src/request/ADAcquireTokenSilentHandler.h [hunk 7]
```diff
@@ -37,6 +37,7 @@
     ADTokenCacheAccessor* _tokenCache;
     
     ADTokenCacheItem* _mrrtItem;
+    ADTokenCacheItem* _validStaleAccessTokenItem; //store valid AT in terms of ext_expires_in (if find any)
```
Comment (heidijinxujia) [inline ADAL/src/request/ADAcquireTokenSilentHandler.h]: _extendedLifetimeAccessToken

File: ADAL/src/request/ADAcquireTokenSilentHandler.m [hunk 8]
```diff
@@ -452,4 +470,45 @@ - (void)tryFRT:(NSString*)familyId completionBlock:(ADAuthenticationCallback)com
...
+        completionBlock(result);
+    };
+    newCompletion = [newCompletion copy];
+    SAFE_ARC_AUTORELEASE(newCompletion);
+    return newCompletion;
+}
+
+- (BOOL) isServerUnavailable:(ADAuthenticationResult *)result
+{
+    if ([[result error] code] == 503 || [[result error] code] == 504)
```
Comment (heidijinxujia) [inline ADAL/src/request/ADAcquireTokenSilentHandler.m]: 500 as well

File: ADAL/src/request/ADAcquireTokenSilentHandler.m [hunk 9]
```diff
@@ -452,4 +470,45 @@ - (void)tryFRT:(NSString*)familyId completionBlock:(ADAuthenticationCallback)com
...
 
+/*
+ This method will add the logic of returning a valid AT in terms of ext_expires_in.
+ The logic is appended to the front of the completion block because returning a valid AT in terms of 
+ ext_expires_in happens in the end of the cache look up sequence.
+ */
+- (ADAuthenticationCallback) enableLogicForExtendedLifetime:(ADAuthenticationCallback)completionBlock
+{
+    // logic is added to return valid stale access token when server is unavailable
+    ADAuthenticationCallback newCompletion = ^(ADAuthenticationResult *result)
```
Comment (heidijinxujia) [inline ADAL/src/request/ADAcquireTokenSilentHandler.m]: see how to add this part of logic into the end of lookup sequence

File: ADAL/src/request/ADAcquireTokenSilentHandler.m [hunk 10]
```diff
@@ -452,4 +470,45 @@ - (void)tryFRT:(NSString*)familyId completionBlock:(ADAuthenticationCallback)com
...
+    {
+        if ([self isServerUnavailable:result] && _validStaleAccessTokenItem)
```
Comment (heidijinxujia) [inline ADAL/src/request/ADAcquireTokenSilentHandler.m]: only if there is error being returned.

File: ADAL/src/cache/ADTokenCacheItem+Internal.m (hunk 3 above)
Comment (jasoncoolmax) [inline ADAL/src/cache/ADTokenCacheItem+Internal.m]: Sure :)

File: ADAL/src/cache/ADTokenCacheItem+Internal.m [hunk 11]
```diff
@@ -196,29 +219,13 @@ - (BOOL)fillItemWithResponse:(NSDictionary*)responseDictionary
     
     [self fillExpiration:responseDictionary];
     
-    BOOL isMRRT = ![NSString adIsStringNilOrBlank:[responseDictionary objectForKey:OAUTH2_RESOURCE]] && ![NSString adIsStringNilOrBlank:self.refreshToken];
+    BOOL isMRRT = ![NSString adIsStringNilOrBlank:self.resource] && ![NSString adIsStringNilOrBlank:self.refreshToken];
```
Comment (RPangrle) [inline ADAL/src/cache/ADTokenCacheItem+Internal.m]: If you have to move up fill expiration then do it, but this change has the potential to cause unexpected side effects.

Comment (jasoncoolmax) [inline ADAL/src/cache/ADTokenCacheItem+Internal.m]: Shall I move this line to line 211 and change it to be
self.resource CHANGED TO BE [responseDictionary objectForKey:OAUTH2_RESOURCE]
self.refreshToken CHANGED TO BE [responseDictionary objectForKey:OAUTH2_REFRESH_TOKEN]?

**Comment by RPangrle** [comment]
:shipit:

**Pull request merged**